### DISCUSSION TITLE
REINDEX is extremely slow

### DISCUSSION THREAD

**Discussion by Kukant** [JanusGraph/janusgraph#3458] [Q&A]
Hi,
I am using index on one of my vertex properties, because I use it for filtering, here is the code for createing index (basically just copy pasted from documentation):
```
graph.tx().rollback() 
mgmt = graph.openManagement()
fqn = mgmt.getPropertyKey('fqn')
mgmt.buildIndex('byFqn', Vertex.class).addKey(fqn).unique().buildCompositeIndex()
mgmt.commit()
//Wait for the index to become available
ManagementSystem.awaitGraphIndexStatus(graph, 'byFqn').call()
//Reindex the existing data
mgmt = graph.openManagement()
mgmt.updateIndex(mgmt.getGraphIndex("byFqn"), SchemaAction.REINDEX).get()
mgmt.commit()
```

My problem is, that the REINDEX action takes ages (usually about 8 hours) for appx. 4 millions of edges.
During the process, the CPU usage stays very low, which is weird (< 1%).

Is this duration normal? Can I somehow speed up this process?

My Janus server runs in Azure cloud on Standard B8ms machine with 32 GB RAM and 8 CPUs. I have set the server to use about 22 GB RAM for heap. I use Berkeley as backend storage.

**Comment by vtslab** [comment]
JanusGraph includes [a mapreduce job for reindexing ](https://docs.janusgraph.org/schema/index-management/index-reindexing/#executing-a-reindex-job-on-mapreduce)to parallellize the operation. Did you try that?

**Reply by Kukant** [reply]
Good idea,  I have not tried that. However I cannot try it, as I am using Berkeley for backend storage and I dont use Hadoop at all. It would be nice to be able to parallelize with plain JanusGraphManagement. I guess that is impossible now?

**Reply by Kukant** [reply]
Even though it is not parallelized, I would expect the reindex process to use 100% of CPU, or I/O or something... I dont understand where the bottleneck is right now.

**Reply by cdegroc** [reply]
In case you're still looking, @li-boxuan added support for parallel reindexing in #2901 (unreleased, part of JanusGraph 1.0).
Otherwise, could the process be limited by the throughput to your embedded BerkeleyDB backend?

**Comment by mad** [comment]
Try to increase PAGE_SIZE to 50k - used as work block size and default value is too small

**Reply by Kukant** [reply]
do you mean the storage.page-size property?

**Reply by Kukant** [reply]
I tried that, and it sadly had no effect on the indexing speed.
